lib/credentials/assume_role_with_web_identity_credentials.js

```diff
@@ -0,0 +1,153 @@
+   */
+  getParamsFromSharedConfig: function getParamsFromSharedConfig() {
+    try {
+      var profiles = {};
+      var profilesFromConfig = {};
+      if (process.env[AWS.util.configOptInEnv]) {
+        profilesFromConfig = AWS.util.iniLoader.loadFrom({
+          isConfig: true,
+          filename: process.env[AWS.util.sharedConfigFileEnv]
+        });
+      }
+      var profilesFromCreds = AWS.util.iniLoader.loadFrom({
+        filename: (process.env[AWS.util.configOptInEnv] && process.env[AWS.util.sharedCredentialsFileEnv])
+      });
+      for (var i = 0, profileNames = Object.keys(profilesFromConfig); i < profileNames.length; i++) {
+        profiles[profileNames[i]] = profilesFromConfig[profileNames[i]];
+      }
+      for (var i = 0, profileNames = Object.keys(profilesFromCreds); i < profileNames.length; i++) {
+        profiles[profileNames[i]] = profilesFromCreds[profileNames[i]];
+      }
+      var profile = profiles[process.env.AWS_PROFILE || AWS.util.defaultProfile] || {};
+
+      if (Object.keys(profile).length === 0) {
+        throw AWS.util.error(
+          new Error('Profile ' + this.profile + ' not found'),
+          { code: 'AssumeRoleWithWebIdentityCredentialsProviderFailure' }
+        );
+      }
+
+      return {
+        envTokenFile: profile['web_identity_token_file'],
+        roleArn: profile['role_arn'],
+        roleSessionName: profile['role_session_name']
+      };
+    } catch (err) {
+      throw err;
+    }
+  },
+
+  /**
+   * Refreshes credentials using {AWS.STS.assumeRoleWithWebIdentity}
+   *
+   * @callback callback function(err)
+   *   Called when the STS service responds (or fails). When
+   *   this callback is called with no error, it means that the credentials
+   *   information has been loaded into the object (as the `accessKeyId`,
+   *   `secretAccessKey`, and `sessionToken` properties).
+   *   @param err [Error] if an error occurred, this value will be filled
+   * @see AWS.Credentials.get
+   */
+  refresh: function refresh(callback) {
+    this.coalesceRefresh(callback || AWS.util.fn.callback);
+  },
+
+  /**
+   * @api private
+   */
+  getRandomSessionName: function getRandomSessionName() {
+    return crypto.randomBytes(64).toString('base64');
+  },
+
+  /**
+   * @api private
+   */
+  load: function load(callback) {
+    var self = this;
+    try {
+      var params = self.getParamsFromEnv();
+      if (!params) {
+        params = self.getParamsFromSharedConfig();
+      }
+      if (params) {
+        var oidcToken = fs.readFileSync(params.envTokenFile, {encoding: 'ascii'});
+        var webIdentityCredentials = new AWS.WebIdentityCredentials({
+          WebIdentityToken: oidcToken,
+          RoleArn: params.roleArn,
+          RoleSessionName: params.roleSessionName || this.getRandomSessionName()
+        });
+        webIdentityCredentials.refresh(function(webErr) {
+          if (!webErr) {
+            self.sts.credentialsFrom(webIdentityCredentials.data, self);
```

Comment:
If you are going to remove the STS client in `AssumeRoleWithWebIdentityCredentials`, you can use the `credentialsFrom` functions provided by `webIdentityCredentials .service`.

Reply:
Resolved in https://github.com/aws/aws-sdk-js/pull/2737/commits/d33de140cc6dab88ae3dd72ea31e4978db3a3474